File: ZDoom/ZDoomNavmesh/Parsers.cs
/*
 * Author: Pollazzon Stefano
 * Project: ZDoom Navmesh builder
 * This module contains a very simple text lexer and
 * a UDMF parser, that is used in reading text maps
 */
using System;
using ZDoomNavmesh;

namespace NavmeshBuilder
{
    /// <summary>
    /// Class <c>TScanner</c> is a simple text scanner, that maintains a position, row and column of the processing chars.
    /// </summary>
    internal sealed class TScanner : Object
    {
        private readonly string TextMap;
        private int FColumn;
        private int FRow;
        private int Position;
        public int Column { get => FColumn; }
        public int Row { get => FRow; }
        public bool Finished { get => Position >= TextMap.Length; }
        public TScanner(string ATextMap)
        {
            TextMap = ATextMap ?? throw new ArgumentNullException(nameof(ATextMap));
            FColumn = 1;
            FRow = 1;
            Position = 1;
        }
        public char GetNextChar()
        {
            char Result;
            if (Position <= TextMap.Length)
            {
                Result = TextMap[Position - 1];
                Position += 1;
                if (Result == '\n')
                {
                    FColumn = 1;
                    FRow += 1;
                }
                else
                    FColumn += 1;
            }
            else
                Result = '\0';
            return Result;
        }
        public char Peek(int Offset = 0)
        {
            char Result;
            if ((Position + Offset >= 1) && (Position + Offset <= TextMap.Length))
            {
                Result = TextMap[Position + Offset - 1];
                if (Result == '\n')
                {
                    FColumn = 1;
                    FRow += 1;
                }
                else
                    FColumn += 1;
            }
            else
                Result = '\0';
            return Result;
        }
    }
    /// <summary>
    /// Class <c>TUdmfLexer</c> is a simple text lexer of the UDMF grammar.
    /// </summary>
    public sealed class TUdmfLexer : Object
    {
        private TScanner Scanner;
        public int Column { get => Scanner.Column; }
        public int Row { get => Scanner.Row; }
        public TUdmfLexer(string TextMap)
        {
            if (TextMap == null)
                throw new ArgumentNullException(nameof(TextMap));
            Scanner = new TScanner(TextMap);
        }
        private char GetNextChar()
        {
            if (Scanner.Finished)
                return '\0';
            else
                return Scanner.GetNextChar();
        }
        private bool IsSymbol(char C)
        {
            return C == ';';
        }
        private bool IsWhitespace(char C)
        {
            return (C == ' ') || (C == '\t') || (C == '\n') || (C == '\r');
        }
        private bool IsTokenStop(char C)
        {
            return (IsSymbol(C)) || (IsWhitespace(C));
        }
        private void SkipWhitespaces()
        {
            while (IsWhitespace(Scanner.Peek(0)))
                if (GetNextChar() == '\0')
                    break;
        }
        /// <summary>
        /// Function <c>GetToken</c> gives the next token in the source code.
        /// </summary>
        /// <returns>
        /// A string representing the next token in the source code, without leading or trailing whitespaces.
        /// </returns>
        public string GetToken()
        {
            bool OpenQuote1 = false;
            bool OpenQuote2 = false;
            if (Scanner.Finished)
                return null;
            string Result = "";
            do
            {
                char C = GetNextChar();
                if ((C == '/') && (!OpenQuote1) && (!OpenQuote2))
                {
                    C = GetNextChar();
                    if (C == '/')
                    {
                        do
                        {
                            C = GetNextChar();
                        } while (C != '\n');
                        SkipWhitespaces();
                    }
                    else
                        if (C == '*')
                    {
                        do
                        {
                            C = GetNextChar();
                        } while ((C != '*') || (Scanner.Peek() != '/'));
                        C = GetNextChar();
                        SkipWhitespaces();
                    }
                }
                else
                    if ((C == '\'') && (!OpenQuote1))
                    OpenQuote1 = false;
                else
                        if ((C == '"') && (!OpenQuote2))
                    OpenQuote2 = false;
                else
                            if ((C == '\\') && ((OpenQuote1) || (OpenQuote2)))
                    C = GetNextChar();
                else
                    Result += C;
                if (Scanner.Finished)
                    break;
            } while ((!IsTokenStop(Scanner.Peek(0))) || (OpenQuote1) || (OpenQuote2));
            SkipWhitespaces();
            return Result;
        }
    }
    /// <summary>
    /// Class <c>EUdmfException</c> is an Exception thrown, when there are problems in parsing the source code.
    /// </summary>
    public class EUdmfException : FormatException
    {
        public EUdmfException()
        {
        }
        public EUdmfException(string Message) : base(Message)
        {
        }
        public EUdmfException(string Message, Exception InnerException) : base(Message, InnerException)
        {
        }
    }
    /// <summary>
    /// Class <c>TUdmfParser</c> is a simple UDMF parser.
    /// </summary>
    public sealed class TUdmfParser : Object
    {
        private TUdmfLexer UdmfLexer;
        private TMapNamespace FMapNamespace;
        public TMapNamespace MapNamespace { get => FMapNamespace; }
        private bool Expects(string Value)
        {
            string Token = UdmfLexer.GetToken();
            return string.Equals(Token, Value, StringComparison.OrdinalIgnoreCase);
        }
        private string GetValueToken()
        {
            string Result;
            if (Expects("="))
            {
                Result = UdmfLexer.GetToken();
                if (!Expects(";"))
                    throw new EUdmfException(string.Format("Row %n, column %n: expected ';'", UdmfLexer.Row, UdmfLexer.Column));
            }
            else
                throw new EUdmfException(string.Format("Row %n, column %n: expected '='", UdmfLexer.Row, UdmfLexer.Column));
            return Result;
        }
        private bool GetBooleanToken()
        {
            string Token = GetValueToken();
            bool Result = Boolean.Parse(Token);
            return Result;
        }
        private short GetInt16Token()
        {
            string Token = GetValueToken();
            short Result = Convert.ToInt16(Double.Parse(Token, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture));
            return Result;
        }
        private int GetInt32Token()
        {
            string Token = GetValueToken();
            int Result = Convert.ToInt32(Double.Parse(Token, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture));
            return Result;
        }
        private TMapVertex GetMapVertex()
        {
            int X = 0;
            int Y = 0;
            bool ValidX = false;
            bool ValidY = false;
            string Token;
            Expects("{");
            do
            {
                Token = UdmfLexer.GetToken();
                if (string.Equals(Token, "X", StringComparison.OrdinalIgnoreCase))
                {
                    X = GetInt32Token();
                    ValidX = true;
                }
                if (string.Equals(Token, "Y", StringComparison.OrdinalIgnoreCase))
                {
                    Y = GetInt32Token();
                    ValidY = true;
                }
            } while (Token != "}");
            if ((ValidX) && (ValidY))
                return new TMapVertex
                {
                    X = X,
                    Y = Y
                };
            else
                return null;
        }
        private TMapLinedef GetMapLinedef()
        {
            int V1 = 0;
            int V2 = 0;
            bool ValidV1 = false;
            bool ValidV2 = false;
            int ID = 0;
            bool Blocking = false;
            bool BlockMonsters = false;
            bool TwoSided = false;
            int SideFront = 0;
            int SideBack = 0;
            short Special = 0;
            short Arg0 = 0;
            short Arg1 = 0;
            short Arg2 = 0;
            short Arg3 = 0;
            short Arg4 = 0;
            bool MonsterUse = false;
            bool RepeatableSpecial = false;
            bool Ignored = false;
            string Token;
            Expects("{");
            do
            {
                Token = UdmfLexer.GetToken();
                if (string.Equals(Token, "V1", StringComparison.OrdinalIgnoreCase))
                {
                    V1 = GetInt32Token();
                    ValidV1 = true;
                }
                if (string.Equals(Token, "V2", StringComparison.OrdinalIgnoreCase))
                {
                    V2 = GetInt32Token();
                    ValidV2 = true;
                }
                if (string.Equals(Token, "ID", StringComparison.OrdinalIgnoreCase))
                    ID = GetInt16Token();
                if (string.Equals(Token, "Blocking", StringComparison.OrdinalIgnoreCase))
                    Blocking = GetBooleanToken();
                if (string.Equals(Token, "BlockMonsters", StringComparison.OrdinalIgnoreCase))
                    BlockMonsters = GetBooleanToken();
                if (string.Equals(Token, "TwoSided", StringComparison.OrdinalIgnoreCase))
                    TwoSided = GetBooleanToken();
                if (string.Equals(Token, "SideFront", StringComparison.OrdinalIgnoreCase))
                    SideFront = GetInt32Token();
                if (string.Equals(Token, "SideBack", StringComparison.OrdinalIgnoreCase))
                    SideBack = GetInt32Token();
                if (string.Equals(Token, "Special", StringComparison.OrdinalIgnoreCase))
                    Special = GetInt16Token();
                if (string.Equals(Token, "Arg0", StringComparison.OrdinalIgnoreCase))
                    Arg0 = GetInt16Token();
                if (string.Equals(Token, "Arg1", StringComparison.OrdinalIgnoreCase))
                    Arg1 = GetInt16Token();
                if (string.Equals(Token, "Arg2", StringComparison.OrdinalIgnoreCase))
                    Arg2 = GetInt16Token();
                if (string.Equals(Token, "Arg3", StringComparison.OrdinalIgnoreCase))
                    Arg3 = GetInt16Token();
                if (string.Equals(Token, "Arg4", StringComparison.OrdinalIgnoreCase))
                    Arg4 = GetInt16Token();
                if (string.Equals(Token, "MonsterUse", StringComparison.OrdinalIgnoreCase))
                    MonsterUse = GetBooleanToken();
                if (string.Equals(Token, "RepeatableSpecial", StringComparison.OrdinalIgnoreCase))
                    RepeatableSpecial = GetBooleanToken();
                if (string.Equals(Token, "user_norecast", StringComparison.OrdinalIgnoreCase))
                    Ignored = GetBooleanToken();
            } while (Token != "}");
            if ((ValidV1) && (ValidV2))
                return new TMapLinedef
                {
                    V1 = V1,
                    V2 = V2,
                    ID = ID,
                    Blocking = Blocking || BlockMonsters,
                    TwoSided = TwoSided,
                    SideFront = SideFront,
                    SideBack = SideBack,
                    Special = Special,
                    Arg0 = Arg0,
                    Arg1 = Arg1,
                    Arg2 = Arg2,
                    Arg3 = Arg3,
                    Arg4 = Arg4,
                    MonsterUse = MonsterUse,
                    RepeatableSpecial = RepeatableSpecial,
                    Ignored = Ignored
                };
            else
                return null;
        }
        private TMapSidedef GetMapSidedef()
        {
            int Sector = 0;
            bool ValidSector = false;
            string Token;
            Expects("{");
            do
            {
                Token = UdmfLexer.GetToken();
                if (string.Equals(Token, "Sector", StringComparison.OrdinalIgnoreCase))
                {
                    Sector = GetInt32Token();
                    ValidSector = true;
                }
            } while (Token != "}");
            if (ValidSector)
                return new TMapSidedef
                {
                    Sector = Sector
                };
            else
                return null;
        }
        private TMapSector GetMapSector()
        {
            int HeightFloor = 0;
            int HeightCeiling = 0;
            int ID = 0;
            short Special = 0;
            bool Ignored = false;
            string Token;
            Expects("{");
            do
            {
                Token = UdmfLexer.GetToken();
                if (string.Equals(Token, "HeightFloor", StringComparison.OrdinalIgnoreCase))
                    HeightFloor = GetInt32Token();
                if (string.Equals(Token, "HeightCeiling", StringComparison.OrdinalIgnoreCase))
                    HeightCeiling = GetInt32Token();
                if (string.Equals(Token, "ID", StringComparison.OrdinalIgnoreCase))
                    ID = GetInt32Token();
                if (string.Equals(Token, "Special", StringComparison.OrdinalIgnoreCase))
                    Special = GetInt16Token();
                if (string.Equals(Token, "user_norecast", StringComparison.OrdinalIgnoreCase))
                    Ignored = GetBooleanToken();
            } while (Token != "}");
            return new TMapSector
            {
                HeightFloor = HeightFloor,
                HeightCeiling = HeightCeiling,
                ID = ID,
                Special = Special,
                Ignored = Ignored
            };
        }
        private TMapThing GetMapThing()
        {
            int X = 0;
            int Y = 0;
            int Z = 0;
            bool ValidX = false;
            bool ValidY = false;
            int ID = 0;
            short Special = 0;
            short Arg0 = 0;
            short Arg1 = 0;
            short Arg2 = 0;
            short Arg3 = 0;
            short Arg4 = 0;
            string Token;
            Expects("{");
            do
            {
                Token = UdmfLexer.GetToken();
                if (string.Equals(Token, "X", StringComparison.OrdinalIgnoreCase))
                {
                    X = GetInt32Token();
                    ValidX = true;
                }
                if (string.Equals(Token, "Y", StringComparison.OrdinalIgnoreCase))
                {
                    X = GetInt32Token();
                    ValidY = true;
                }
                if (string.Equals(Token, "Z", StringComparison.OrdinalIgnoreCase))
                    Z = GetInt32Token();
                if (string.Equals(Token, "ID", StringComparison.OrdinalIgnoreCase))
                    ID = GetInt32Token();
                if (string.Equals(Token, "Special", StringComparison.OrdinalIgnoreCase))
                    Special = GetInt16Token();
                if (string.Equals(Token, "Arg0", StringComparison.OrdinalIgnoreCase))
                    Arg0 = GetInt16Token();
                if (string.Equals(Token, "Arg1", StringComparison.OrdinalIgnoreCase))
                    Arg1 = GetInt16Token();
                if (string.Equals(Token, "Arg2", StringComparison.OrdinalIgnoreCase))
                    Arg2 = GetInt16Token();
                if (string.Equals(Token, "Arg3", StringComparison.OrdinalIgnoreCase))
                    Arg3 = GetInt16Token();
                if (string.Equals(Token, "Arg4", StringComparison.OrdinalIgnoreCase))
                    Arg4 = GetInt16Token();

            } while (Token != "}");
            if ((ValidX) && (ValidY))
                return new TMapThing
                {
                    X = X,
                    Y = Y,
                    Z = Z,
                    ID = ID,
                    Special = Special,
                    Arg0 = Arg0,
                    Arg1 = Arg1,
                    Arg2 = Arg2,
                    Arg3 = Arg3,
                    Arg4 = Arg4
                };
            else
                return null;
        }
        public TUdmfParser(string TextMap)
        {
            if (TextMap is null)
                throw new ArgumentNullException(nameof(TextMap));
            UdmfLexer = new TUdmfLexer(TextMap);
            string Token = UdmfLexer.GetToken();
            if (string.Equals(Token, "NAMESPACE", StringComparison.OrdinalIgnoreCase))
            {
                string TextNamespace = GetValueToken();
                if (string.Equals(TextNamespace, "DOOM", StringComparison.OrdinalIgnoreCase))
                    FMapNamespace = TMapNamespace.MapNamespaceDoom;
                else
                    if (string.Equals(TextNamespace, "HEXEN", StringComparison.OrdinalIgnoreCase))
                        FMapNamespace = TMapNamespace.MapNamespaceHexen;
                    else
                        if (string.Equals(TextNamespace, "ZDOOM", StringComparison.OrdinalIgnoreCase))
                            FMapNamespace = TMapNamespace.MapNamespaceZDoom;
                        else
                            throw new EUdmfException(string.Format("Row %n, column %n: invalid namespace '%s'", UdmfLexer.Row, UdmfLexer.Column, Token));
            }
            else
                throw new EUdmfException(string.Format("Row %n, column %n: expected namespace, found '%s'", UdmfLexer.Row, UdmfLexer.Column, Token));
        }
        /// <summary>
        /// Function <c>ParseMap</c> gives the next token in the source code.
        /// </summary>
        /// <param name="MapDefinition"><c>MapDefinition</c> is the map, to fill with the TEXTMAP source code.</param>
        public void ParseMap(TMapDefinition MapDefinition)
        {
            if (MapDefinition == null)
                throw new ArgumentNullException(nameof(MapDefinition));
            string Token = UdmfLexer.GetToken();
            while (!string.IsNullOrEmpty(Token))
            {
                if (string.Equals(Token, "VERTEX", StringComparison.OrdinalIgnoreCase))
                    MapDefinition.AddMapVertex(GetMapVertex());
                else
                    if (string.Equals(Token, "LINEDEF", StringComparison.OrdinalIgnoreCase))
                        MapDefinition.AddMapLinedef(GetMapLinedef());
                    else
                        if (string.Equals(Token, "SIDEDEF", StringComparison.OrdinalIgnoreCase))
                            MapDefinition.AddMapSidedef(GetMapSidedef());
                        else
                            if (string.Equals(Token, "SECTOR", StringComparison.OrdinalIgnoreCase))
                                MapDefinition.AddMapSector(GetMapSector());
                            else
                                if (string.Equals(Token, "THING", StringComparison.OrdinalIgnoreCase))
                                    MapDefinition.AddMapThing(GetMapThing());
                                else
                                    throw new EUdmfException(string.Format("Row %n, column %n: unknown UDMF entity '%s'", UdmfLexer.Row, UdmfLexer.Column, Token));
                Token = UdmfLexer.GetToken();
            }
        }
    }
}

File: Generator/Python/Parsers.py
"""
- Author: Pollazzon Stefano (ported)
- Project: ZDoom Navmesh builder
- This module contains a very simple text lexer and
- a UDMF parser, that is used in reading text maps
"""

from Structures import (
    TMapVertex,
    TMapLinedef,
    TMapSidedef,
    TMapSector,
    TMapThing,
    TMapNamespace,
)


class TScanner:
    """
    Simple text scanner that maintains position, row, and column.
    """

    def __init__(self, TextMap):
        if TextMap is None:
            raise ValueError("TextMap cannot be None")
        self.TextMap = TextMap
        self._column = 1
        self._row = 1
        self._position = 1  # 1-based, like the C# code

    @property
    def Column(self):
        return self._column

    @property
    def Row(self):
        return self._row

    @property
    def Finished(self):
        return self._position > len(self.TextMap)

    def GetNextChar(self):
        if self._position <= len(self.TextMap):
            result = self.TextMap[self._position - 1]
            self._position += 1
            if result == "\n":
                self._column = 1
                self._row += 1
            else:
                self._column += 1
            return result
        return "\0"

    def Peek(self, Offset=0):
        pos = self._position + Offset
        if 1 <= pos <= len(self.TextMap):
            return self.TextMap[pos - 1]
        return "\0"


class TUdmfLexer:
    """
    Simple lexer for UDMF text.
    """

    def __init__(self, TextMap):
        if TextMap is None:
            raise ValueError("TextMap cannot be null")
        self.Scanner = TScanner(TextMap)

    @property
    def Column(self):
        return self.Scanner.Column

    @property
    def Row(self):
        return self.Scanner.Row

    def GetNextChar(self):
        return "\0" if self.Scanner.Finished else self.Scanner.GetNextChar()

    def IsSymbol(self, C):
        # More robust than the original while fully compatible with your UDMF files
        return C in ";{}="

    def IsWhitespace(self, C):
        return C in (" ", "\t", "\n", "\r")

    def IsTokenStop(self, C):
        return self.IsSymbol(C) or self.IsWhitespace(C)

    def SkipWhitespaces(self):
        while True:
            c = self.Scanner.Peek(0)

            # normal whitespace
            if self.IsWhitespace(c):
                if self.GetNextChar() == "\0":
                    break
                continue

            # single-line comments // ...
            if c == "/" and self.Scanner.Peek(1) == "/":
                self.GetNextChar()
                self.GetNextChar()
                while True:
                    c = self.GetNextChar()
                    if c in ("\0", "\n"):
                        break
                continue

            # block comments /* ... */
            if c == "/" and self.Scanner.Peek(1) == "*":
                self.GetNextChar()
                self.GetNextChar()
                while True:
                    c = self.GetNextChar()
                    if c == "\0":
                        break
                    if c == "*" and self.Scanner.Peek(0) == "/":
                        self.GetNextChar()
                        break
                continue

            break

    def GetToken(self):
        self.SkipWhitespaces()

        if self.Scanner.Finished:
            return ""

        c = self.Scanner.Peek(0)

        # single-symbol token
        if self.IsSymbol(c):
            self.GetNextChar()
            self.SkipWhitespaces()
            return c

        result = []
        open_quote1 = False
        open_quote2 = False

        while True:
            c = self.GetNextChar()
            if c == "\0":
                break

            if not open_quote1 and not open_quote2 and c == "/" and self.Scanner.Peek(0) == "/":
                # consume second slash
                self.GetNextChar()
                while True:
                    c = self.GetNextChar()
                    if c in ("\0", "\n"):
                        break
                break

            if not open_quote1 and not open_quote2 and c == "/" and self.Scanner.Peek(0) == "*":
                self.GetNextChar()
                while True:
                    c = self.GetNextChar()
                    if c == "\0":
                        break
                    if c == "*" and self.Scanner.Peek(0) == "/":
                        self.GetNextChar()
                        break
                break

            if c == "'" and not open_quote2:
                open_quote1 = not open_quote1
                continue

            if c == '"' and not open_quote1:
                open_quote2 = not open_quote2
                continue

            if c == "\\" and (open_quote1 or open_quote2):
                c = self.GetNextChar()
                if c == "\0":
                    break
                result.append(c)
                continue

            if (not open_quote1 and not open_quote2):
                if self.IsSymbol(c):
                    # put back symbol by rewinding one character
                    self.Scanner._position -= 1
                    if self.Scanner._column > 1:
                        self.Scanner._column -= 1
                    break
                if self.IsWhitespace(c):
                    break

            result.append(c)

        self.SkipWhitespaces()
        return "".join(result)


class EUdmfException(Exception):
    pass


class TUdmfParser:
    def __init__(self, TextMap):
        if TextMap is None:
            raise ValueError("TextMap cannot be null")

        self.UdmfLexer = TUdmfLexer(TextMap)

        token = self.UdmfLexer.GetToken()
        if token.lower() == "namespace":
            text_namespace = self.GetValueToken()
            if text_namespace.lower() == "doom":
                self.MapNamespace = TMapNamespace.MapNamespaceDoom
            elif text_namespace.lower() == "hexen":
                self.MapNamespace = TMapNamespace.MapNamespaceHexen
            elif text_namespace.lower() == "zdoom":
                self.MapNamespace = TMapNamespace.MapNamespaceZDoom
            else:
                raise EUdmfException(
                    f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: invalid namespace '{text_namespace}'"
                )
        else:
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected namespace, found '{token}'"
            )

    # --------------------------------------------------
    # helpers
    # --------------------------------------------------

    def Expects(self, Value):
        token = self.UdmfLexer.GetToken()
        return token.lower() == Value.lower()

    def GetValueToken(self):
        if self.Expects("="):
            result = self.UdmfLexer.GetToken()
            if not self.Expects(";"):
                raise EUdmfException(
                    f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected ';'"
                )
            return result
        raise EUdmfException(
            f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '='"
        )

    def GetBooleanToken(self):
        token = self.GetValueToken().strip().lower()
        if token == "true":
            return True
        if token == "false":
            return False
        raise EUdmfException(
            f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: invalid boolean '{token}'"
        )

    def GetInt16Token(self):
        token = self.GetValueToken()
        return int(round(float(token)))

    def GetInt32Token(self):
        token = self.GetValueToken()
        return int(round(float(token)))

    # --------------------------------------------------
    # entity readers
    # --------------------------------------------------

    def GetMapVertex(self):
        x = 0
        y = 0
        valid_x = False
        valid_y = False

        if not self.Expects("{"):
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '{{'"
            )

        token = self.UdmfLexer.GetToken()
        while token != "}":
            if token.lower() == "x":
                x = self.GetInt32Token()
                valid_x = True
            elif token.lower() == "y":
                y = self.GetInt32Token()
                valid_y = True
            else:
                # ignore unsupported fields
                if self.UdmfLexer.Scanner.Peek(0) == "=":
                    self.GetValueToken()
            token = self.UdmfLexer.GetToken()

        if valid_x and valid_y:
            return TMapVertex(x, y)
        return None

    def GetMapLinedef(self):
        v1 = 0
        v2 = 0
        valid_v1 = False
        valid_v2 = False

        line_id = 0
        blocking = False
        block_monsters = False
        two_sided = False
        side_front = 0
        side_back = 0
        special = 0
        arg0 = 0
        arg1 = 0
        arg2 = 0
        arg3 = 0
        arg4 = 0
        monster_use = False
        repeatable_special = False
        ignored = False

        if not self.Expects("{"):
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '{{'"
            )

        token = self.UdmfLexer.GetToken()
        while token != "}":
            t = token.lower()

            if t == "v1":
                v1 = self.GetInt32Token()
                valid_v1 = True
            elif t == "v2":
                v2 = self.GetInt32Token()
                valid_v2 = True
            elif t == "id":
                line_id = self.GetInt16Token()
            elif t == "blocking":
                blocking = self.GetBooleanToken()
            elif t == "blockmonsters":
                block_monsters = self.GetBooleanToken()
            elif t == "twosided":
                two_sided = self.GetBooleanToken()
            elif t == "sidefront":
                side_front = self.GetInt32Token()
            elif t == "sideback":
                side_back = self.GetInt32Token()
            elif t == "special":
                special = self.GetInt16Token()
            elif t == "arg0":
                arg0 = self.GetInt16Token()
            elif t == "arg1":
                arg1 = self.GetInt16Token()
            elif t == "arg2":
                arg2 = self.GetInt16Token()
            elif t == "arg3":
                arg3 = self.GetInt16Token()
            elif t == "arg4":
                arg4 = self.GetInt16Token()
            elif t == "monsteruse":
                monster_use = self.GetBooleanToken()
            elif t == "repeatablespecial":
                repeatable_special = self.GetBooleanToken()
            elif t in ("user_norecast", "user_nocast"):
                ignored = self.GetBooleanToken()
            else:
                if self.UdmfLexer.Scanner.Peek(0) == "=":
                    self.GetValueToken()

            token = self.UdmfLexer.GetToken()

        if valid_v1 and valid_v2:
            return TMapLinedef(
                v1,
                v2,
                line_id,
                blocking or block_monsters,
                two_sided,
                side_front,
                side_back,
                special,
                arg0,
                arg1,
                arg2,
                arg3,
                arg4,
                monster_use,
                repeatable_special,
                ignored,
            )
        return None

    def GetMapSidedef(self):
        sector = 0
        valid_sector = False

        if not self.Expects("{"):
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '{{'"
            )

        token = self.UdmfLexer.GetToken()
        while token != "}":
            if token.lower() == "sector":
                sector = self.GetInt32Token()
                valid_sector = True
            else:
                if self.UdmfLexer.Scanner.Peek(0) == "=":
                    self.GetValueToken()
            token = self.UdmfLexer.GetToken()

        if valid_sector:
            return TMapSidedef(sector)
        return None

    def GetMapSector(self):
        height_floor = 0
        height_ceiling = 0
        sector_id = 0
        special = 0
        ignored = False

        if not self.Expects("{"):
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '{{'"
            )

        token = self.UdmfLexer.GetToken()
        while token != "}":
            t = token.lower()

            if t == "heightfloor":
                height_floor = self.GetInt32Token()
            elif t == "heightceiling":
                height_ceiling = self.GetInt32Token()
            elif t == "id":
                sector_id = self.GetInt32Token()
            elif t == "special":
                special = self.GetInt16Token()
            elif t in ("user_norecast", "user_nocast"):
                ignored = self.GetBooleanToken()
            else:
                if self.UdmfLexer.Scanner.Peek(0) == "=":
                    self.GetValueToken()

            token = self.UdmfLexer.GetToken()

        return TMapSector(height_floor, height_ceiling, sector_id, special, ignored)

    def GetMapThing(self):
        x = 0
        y = 0
        z = 0
        thing_type = 0
        valid_x = False
        valid_y = False
        valid_thing_type = False

        thing_id = 0
        special = 0
        arg0 = 0
        arg1 = 0
        arg2 = 0
        arg3 = 0
        arg4 = 0

        if not self.Expects("{"):
            raise EUdmfException(
                f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: expected '{{'"
            )

        token = self.UdmfLexer.GetToken()
        while token != "}":
            t = token.lower()

            if t == "x":
                x = self.GetInt32Token()
                valid_x = True
            elif t == "y":
                y = self.GetInt32Token()
                valid_y = True
            elif t == "type":
                thing_type = self.GetInt32Token()
                valid_thing_type = True
            elif t == "z":
                z = self.GetInt32Token()
            elif t == "id":
                thing_id = self.GetInt32Token()
            elif t == "special":
                special = self.GetInt16Token()
            elif t == "arg0":
                arg0 = self.GetInt16Token()
            elif t == "arg1":
                arg1 = self.GetInt16Token()
            elif t == "arg2":
                arg2 = self.GetInt16Token()
            elif t == "arg3":
                arg3 = self.GetInt16Token()
            elif t == "arg4":
                arg4 = self.GetInt16Token()
            else:
                if self.UdmfLexer.Scanner.Peek(0) == "=":
                    self.GetValueToken()

            token = self.UdmfLexer.GetToken()

        if valid_x and valid_y and valid_thing_type:
            return TMapThing(x, y, z, thing_type, thing_id, special, arg0, arg1, arg2, arg3, arg4)
        return None

    # --------------------------------------------------
    # whole map
    # --------------------------------------------------

    def ParseMap(self, MapDefinition):
        if MapDefinition is None:
            raise ValueError("MapDefinition cannot be None")

        token = self.UdmfLexer.GetToken()
        while token:
            t = token.lower()

            if t == "vertex":
                MapDefinition.AddMapVertex(self.GetMapVertex())
            elif t == "linedef":
                MapDefinition.AddMapLinedef(self.GetMapLinedef())
            elif t == "sidedef":
                MapDefinition.AddMapSidedef(self.GetMapSidedef())
            elif t == "sector":
                MapDefinition.AddMapSector(self.GetMapSector())
            elif t == "thing":
                MapDefinition.AddMapThing(self.GetMapThing())
            else:
                raise EUdmfException(
                    f"Row {self.UdmfLexer.Row}, column {self.UdmfLexer.Column}: unknown UDMF entity '{token}'"
                )

            token = self.UdmfLexer.GetToken()
